database/src/ops/block.rs

```diff
@@ -1,89 +1,422 @@
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    add_block_inner(&mut env.open_tables_mut(tx_rw)?, block)
 }
 
 /// TODO
-pub fn add_block_data() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &Rw,
+    blocks: &[VerifiedBlockInformation],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut tables = env.open_tables_mut(tx_rw)?;
+    for block in blocks {
+        add_block_inner(&mut tables, block)?;
+    }
+    Ok(())
+}
+
+/// Internal function that [`add_block()`] and [`add_blocks()`] uses.
+///
+/// It takes the already opened tables.
+#[allow(clippy::cast_possible_truncation)] // TODO: remove me
+fn add_block_inner(
+    tables: &mut impl TablesMut,
+    block: &VerifiedBlockInformation,
+) -> Result<(), RuntimeError> {
+    // Branch on the hard fork version (`major_version`)
+    // and add the block to the appropriate table.
+    // <https://monero-book.cuprate.org/consensus_rules/hardforks.html#Mainnet-Hard-Forks>
+    //
+    // FIXME: use `match` with ranges when stable:
+    // <https://github.com/rust-lang/rust/issues/37854>
+    if block.block.header.major_version < 4 {
+        tables.block_info_v1s_mut().put(
+            &block.height,
+            &BlockInfoV1 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+            },
+        )
+    } else if block.block.header.major_version < 10 {
+        tables.block_info_v2s_mut().put(
+            &block.height,
+            &BlockInfoV2 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(), // TODO
+            },
+        )
+    } else {
+        tables.block_info_v3s_mut().put(
+            &block.height,
+            &BlockInfoV3 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty,
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(),                    // TODO
+                long_term_weight: block.long_term_weight as u64, // TODO
+            },
+        )
+    }?;
+
+    // BlockBlobs: BlockHeight => BlockBlob
+    // TODO: what is a block blob in Cuprate's case?
+    tables.block_blobs_mut().put(&block.height, todo!())?;
```

Comment:
Btw, this is what the usage looks like.

`tables` here is an `impl TablesMut`, we're getting access to `&mut impl DatabaseRw<BlockBlobs>` via `TablesMut::block_blobs_mut()` here then calling `put()`.